include/readout/models/TaskRawDataProcessorModel.hpp

```diff
@@ -36,37 +39,105 @@ class TaskRawDataProcessorModel : public RawDataProcessorConcept<ReadoutType>
+                                          this,
+                                          std::ref(m_post_process_functions[i]),
+                                          std::ref(*m_producer_consumer_queues[i]));
+    }
+  }
+
+  void stop(const nlohmann::json& /*args*/) override
+  {
+    m_run_marker.store(false);
+    for (auto& thread : m_post_process_threads) {
+      while (!thread->get_readiness()) {
+        std::this_thread::sleep_for(std::chrono::milliseconds(50));
+      }
+    }
+  }
+
+  virtual void get_info(datalinkhandlerinfo::Info& /*info*/)
+  {
+    // No stats for now, extend later
+  }
+
+  void preprocess_item(ReadoutType* item) override { invoke_all_preprocess_functions(item); }
+
+  void postprocess_item(const ReadoutType* item) override
+  {
+    for (size_t i = 0; i < m_producer_consumer_queues.size(); ++i) {
+      if (!m_producer_consumer_queues[i]->write(item)) {
+        ers::warning(PostprocessingNotKeepingUp(ERS_HERE, i));
+      }
+    }
+  }
 
-  void process_item(ReadoutType* item) { invoke_all(item); }
+  template<typename Task>
+  void add_preprocess_task(Task&& task)
+  {
+    m_preprocess_functions.push_back(std::forward<Task>(task));
+  }
 
   template<typename Task>
-  void add_task(Task&& task)
+  void add_postprocess_task(Task&& task)
   {
-    m_tasklist.push_back(std::forward<Task>(task));
+    m_post_process_threads.emplace_back(std::make_unique<ReusableThread>(0));
+    m_post_process_threads.back()->set_name("postprocess", m_post_process_threads.size() - 1);
+    m_post_process_functions.push_back(std::forward<Task>(task));
   }
 
-  void invoke_all(ReadoutType* item)
+  void invoke_all_preprocess_functions(ReadoutType* item)
   {
-    for (auto&& task : m_tasklist) {
+    for (auto&& task : m_preprocess_functions) {
       task(item);
     }
   }
 
   void launch_all(ReadoutType* item)
   {
-    for (auto&& task : m_tasklist) {
+    for (auto&& task : m_preprocess_functions) {
       auto fut = std::async(std::launch::async, task, item);
     }
   }
 
 protected:
+  void run_post_processing_thread(std::function<void(const ReadoutType*)>& function,
+                                  folly::ProducerConsumerQueue<const ReadoutType*>& queue)
+  {
+    while (m_run_marker.load() || queue.sizeGuess() > 0) {
+      const ReadoutType* element;
+      if (queue.read(element)) {
+        function(element);
+      }
+    }
+  }
+
+  std::atomic<bool> m_run_marker{ false };
   // Async tasks and
-  std::vector<std::function<void(ReadoutType*)>> m_tasklist;
+  std::vector<std::function<void(ReadoutType*)>> m_preprocess_functions;
   std::unique_ptr<FrameErrorRegistry>& m_error_registry;
-  // std::map<std::string, std::function<void(ReadoutType*)>> m_tasklist; // futures
+
+  std::vector<std::function<void(const ReadoutType*)>> m_post_process_functions;
+  std::vector<std::unique_ptr<folly::ProducerConsumerQueue<const ReadoutType*>>> m_producer_consumer_queues;
```

Comment:
Just nitpicking, but could we rename this to something more fitting? (e.g.: m_items_to_process)